### ISSUE TITLE
[Bug/Feature] Fix player search

### ISSUE THREAD

**Issue by jiangtyd** [ripgarpr/garpr#187]
Player search currently requests all players in the db, and filters them client-side. This is unacceptably slow nowadays (which is why I'm marking this as a bug.)

There are a number of ways in which this can be improved:
1. Make an ajax request based on what the user has typed so far. This was bad because we hammered the server with too many requests, so we can debounce the requests (see https://github.com/ripgarpr/garpr/issues/174 from many months ago.)
2. We can continue to cache the list of players that have been loaded on the client side as well.
    2a. It seems like currently, when we make a POST request, the client-side cache of players is cleared and we have to fetch it again; we could try to persist the cache across POSTs as well (with perhaps some invalidation mechanism for the changed players?) I don't feel too worried about the consistency of player names.

**Comment by jiangtyd** [comment]
Re: caching: it may be better/cleaner to implement server-side caching (Redis?) rather than client-side caching, since we can then also cache things like tournaments, rankings, etc., and don't need complicated client code.

**Comment by jschnei** [comment]
Yeah, server-side caching with redis is something I've wanted to do for a while. If we do it well enough and add pagination to the client, I think it is probably possible to avoid caching data on the client at all.

Another approach that is probably worth looking at is indexing mongodb and writing better queries in general; currently, for many queries (and in particular, the player search query), python just loads all players/tournaments/whatever from mongo and then filters over them in python. This was probably fine when there were barely any tournaments, but currently we're storing ~20000 players and each call to get all players moves something like 5 MB from the db to python. This isn't so bad by itself, but several people making several calls like this per second (by typing in the lookahead) can lead to OOM issues (and it definitely did when our server only had 1 GB of RAM; back then you could singlehandedly crash the server by typing fast enough in the lookahead -- I've since upgraded to 4 GB, but this could still be a problem if we don't optimize at all).

For player search, we'd probably want to use Mongo's text indices (https://docs.mongodb.com/manual/core/index-text/), but I haven't looked into this at all, so I'm not sure if it can do what we need it to.

**Comment by BrandonCookeDev** [comment]
Made ticket #211 for the above caching conversation
